### PR TITLE
hub push/pull

### PR DESCRIPTION
**Pull request by efriis**
Description: Adds push/pull functions to interact with the hub
Issue: n/a
Dependencies: `langchainhub`

### CODE REVIEW

File: libs/langchain/langchain/hub.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    Pushes an object to the hub and returns the URL.
+    """
+    client = _get_client(kwargs)
+    manifest_json = dumps(object)
+    resp = client.push(repo_full_name, parent_commit_hash, manifest_json)
+    commit_hash: str = resp["commit"]["commit_hash"]
+    return commit_hash
+
+
+def pull(repo_full_name: str, commit_hash: str, **kwargs: Dict[str, Any]) -> Any:
```
Comment (hwchase17) [inline libs/langchain/langchain/hub.py:38]: is commit hash needed?

File: libs/langchain/langchain/hub.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    # use default api key as fallback
+    if api_key is None:
+        api_key = get_from_env("api_key", "LANGCHAIN_API_KEY")
+    return Client(api_url, api_key=api_key)
+
+
+def push(
+    repo_full_name: str,
+    parent_commit_hash: Optional[str],
```
Comment (hwchase17) [inline libs/langchain/langchain/hub.py:24]: is parent commit hash needed?

File: libs/langchain/langchain/hub.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+def _get_client(kwargs: Dict[str, Any]) -> Client:
+    api_url = get_from_dict_or_env(kwargs, "api_url", "LANGCHAIN_HUB_API_URL")
+
+    # read hub-specific api key
+    api_key = get_from_dict_or_env(kwargs, "api_key", "LANGCHAIN_HUB_API_KEY", None)
+
+    # use default api key as fallback
+    if api_key is None:
+        api_key = get_from_env("api_key", "LANGCHAIN_API_KEY")
+    return Client(api_url, api_key=api_key)
```
Comment (baskaryan) [inline libs/langchain/langchain/hub.py:19]: if you default to None (L14) think it'll still raise an error. also slight pref to explicitly type params

```suggestion
def _get_client(api_url: Optional[str]=None, api_key: Optional[str]=None) -> Client:
    api_url = api_url or get_from_env("api_url", "LANGCHAIN_HUB_API_URL")
    api_key = api_key or get_from_env("api_key", "LANGCHAIN_HUB_API_KEY", default="")
    api_key = api_key or get_from_env("api_key", "LANGCHAIN_API_KEY")
    return Client(api_url, api_key=api_key)
```

File: libs/langchain/pyproject.toml [hunk 4]
```diff
@@ -134,6 +134,7 @@ amazon-textract-caller = {version = "<2", optional = true}
 xata = {version = "^1.0.0a7", optional = true}
 xmltodict = {version = "^0.13.0", optional = true}
 betabageldb = {version = "0.2.32", optional = true, python = ">=3.7,<3.11"}
+langchainhub = "^0.1.1"
```
Comment (hinthornw) [inline libs/langchain/pyproject.toml:137]: Should this be optional?

File: libs/langchain/langchain/hub.py (hunk 2 above)
Comment (efriis) [reply libs/langchain/langchain/hub.py:24]: It's a tradeoff between forcing people to be explicit about which changes come after each other and straightforward being able to immediately "push."

The situation having it avoids is 2 people pushing to the same repo and one of them silently overwriting the other one. In this case, whoever pushes second gets an error (at least while we're doing linear versioning).

I'll put up a doc on this. @nfcampos has some thoughts too.

File: libs/langchain/langchain/hub.py (hunk 1 above)
Comment (efriis) [reply libs/langchain/langchain/hub.py:38]: Similar to above, starting with forcing the interface to be explicit and will consider relaxing the requirement in the doc.

File: libs/langchain/langchain/hub.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
+from typing import Any, Dict, Optional
+
+from langchainhub import Client
```
Comment (baskaryan) [inline libs/langchain/langchain/hub.py:3]: if this is optional dependency probably better to import within relevant functions (ie within _get_client). makes it safer to import code from this module

Comment (efriis) [reply libs/langchain/langchain/hub.py]: Done! Would love a quick check of the type checking bit

**Pull request merged**